File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n")

const [n, k] = input[0].split(" ").map(Number)
const MAX = 100_000
const dp = Array.from({ length: MAX + 1 }, () => -1)

const queue = [n]
dp[n] = 0
let index = 0

while (queue.length > index) {
  const cur = queue[index++]
  if (cur == k) continue

  for (let i of [cur - 1, cur + 1, cur * 2]) {
    if (i >= 0 && i <= MAX && dp[i] == -1) {
      dp[i] = dp[cur] + 1
      queue.push(i)
    }
  }
}

console.log(dp[k])

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
from collections import deque
import sys


def bfs(start, end):
    queue = deque([])
    queue.append(start)
    while queue:
        x = queue.popleft()
        dx = [x-1, x+1, x*2]
        if x == end:
            return dist[end]
        for i in dx:
            if 0 <= i <= MAX and dist[i] == 0:
                dist[i] = dist[x]+1
                queue.append(i)


n, k = map(int, sys.stdin.readline().split())
MAX = 10**5
dist = [0]*(MAX+1)
print(bfs(n, k))
